cpu/stm32l1/periph/pm.c

```diff
@@ -0,0 +1,359 @@
+               for (p = 0; p < 16; p ++) {
+                   /* exclude GPIOs we previously set with pin_set */
+                   if ((pm_portmask_system[i] | pm_portmask_user[i]) & (1 << p)) {
+                       mask &= ~(0x03 << (p*2));
+                   } else {
+                       if ((EXTI->IMR & (1 << p)) && ((((SYSCFG->EXTICR[p >> 2]) >> ((p & 0x03) * 4)) & 0xF) == i)) {
+                           mask &= ~(0x03 << (p*2));
+                       }
+                   }
+               }
+           }
+
+           /* disable pull-ups on GPIOs */
+           tmpreg = port->PUPDR;
+           tmpreg &= ~mask;
+           port->PUPDR = tmpreg;
+
+           /* set GPIOs to AIN mode */
+           tmpreg = port->MODER;
+           tmpreg |= mask;
+           port->MODER = tmpreg;
+       }
+
+    /* restore GPIO clocks */
+    tmpreg = RCC->AHBENR;
+    tmpreg &= ~((uint32_t)0xFF);
+    tmpreg |= ahb_gpio_clocks;
+    periph_clk_en(AHB, tmpreg);
+}
+
+/* restore GPIO settings */
+static void pm_when_i_wake_up (void) {
+    /* enable all GPIO clocks */
+    periph_clk_en(AHB, 0xFF);
+
+    uint8_t i;
+    GPIO_TypeDef *port;
+
+    /* restore GPIO settings */
+    for (i = 0; i < CPU_NUMBER_OF_PORTS; i++) {
+        port = (GPIO_TypeDef *)(GPIOA_BASE + i*(GPIOB_BASE - GPIOA_BASE));
+
+        port->PUPDR = pm_gpio_pupdr[i];
+        port->OTYPER = pm_gpio_otyper[i];
+        port->ODR = pm_gpio_odr[i];
+        port->MODER = pm_gpio_moder[i];
+    }
+
+    /* restore GPIO clocks */
+    tmpreg = RCC->AHBENR;
+    tmpreg &= ~((uint32_t)0xFF);
+    tmpreg |= ahb_gpio_clocks;
+    periph_clk_en(AHB, tmpreg);
+
+}
+
+/* Select CPU clocking between default (PM_ON) and medium-speed (PM_IDLE) */
+static void pm_run_mode(void) {
+    DEBUG("Switching to PM_ON");
+    stmclk_init_sysclk();
+
+#ifdef MODULE_XTIMER
+    /* Recalculate xtimer frequency */
+    /* NB: default XTIMER_HZ clock is 1 MHz, so CPU clock must be at least 1 MHz for xtimer to work properly */
+    xtimer_init();
+#endif
+
+    // /* Recalculate stdio UART baudrate */
+    uart_set_baudrate(UART_STDIO_DEV, UART_STDIO_BAUDRATE);
+}
+
+void pm_set(unsigned mode)
+{
+    unsigned state;
+
+    switch (mode) {
+        case PM_OFF:             /* Stand by mode */
+            DEBUG("Switching to PM_OFF");
+            /* Clear Wakeup flag */
+            PWR->CR |= PWR_CR_CWUF;
+
+            /* Select STANDBY mode */
+            PWR->CR |= PWR_CR_PDDS;
+
+            /* Set SLEEPDEEP bit of Cortex System Control Register */
+            SCB->SCR |= (uint32_t)SCB_SCR_SLEEPDEEP;
+
+            /* Enable Ultra Low Power mode */
+            PWR->CR |= PWR_CR_ULP;
+
+            /* This option is used to ensure that store operations are completed */
+            #if defined (__CC_ARM)
+            __force_stores();
+            #endif
+
+            irq_disable();
+
+            /* Request Wait For Interrupt */
+            __DSB();
+            __WFI();
```

Comment:
Some parts of the code above are already implemented on `cortexm_sleep` function. Please remove the duplicates and call the function instead. The cases below too.

Reply:
In cortexm_sleep it is followed by irq_restore(state);

As a rule, you want to restore core and peripheral clocks before enabling and processing any interrupts, especially when waking up from deep sleep mode with its 65 kHz core clock. So either clock switching should be added to cortexm_sleep or irq_restore removed from it.

Reply:
I have the same issue as stated by @olegart

Reply:
You can still do it when the function `cortexm_sleep` returns, as other implementations do if I'm not mistaken. However I think there is a problem using `irq_restore` since on current master is broken for nucleo-l152. My thoughts were that lot's of things were not implemented (power modes, clocking, etc.) so the problem could be anywhere.

I'll try to make more testing on this and figure out where it breaks and see if we can still reuse the code from cortex_common, which is the ideal.

Reply:
CPU was woken from sleep by some interrupt and it will enter IRQ routine immediately after irq_restore, and with current cortexm_sleep it will happen _before_ clocks are restored.

If it was STOP mode, default clock after it is 4 MHz MSI, so probably you wouldn't notice some slowdown inside IRQ processing, but when using DEEP SLEEP, core clock after wakeup may be as low as 65 kHz (it should be set before entering DEEP SLEEP and it affects sleep mode power consumption).

Been there, done that :) The clock must be restored before re-enabling interrupts.

If you get a HardFault right after __WFI(), it's a problem of clock initialization routine (clk_init), we had that problem with RIOT some time ago — somewhere around end of 2016, I believe, and then we rewrote clk_init for our needs,  all the sequence strictly according to RM0038, and the problem was gone.

Reply:
Yes. Just tested it and I found that. Interestingly the problem is not present on other devices, and I was also wondering if the missing configuration of FAULTMASK (see [comment](https://github.com/RIOT-OS/RIOT/issues/8024#issuecomment-356995201)) was also in the origin of the problem.

Thus, I agree to keep this code, but maybe we can factor out the calls to `__DSB()` and `__WFI()` which belong to the Cortex-M architecture and are independent of the manufacturer (ST here).

Reply:
Actually, in `cortexm_sleep`, the `irq_restore` call enables irq only if `pm_set` was called with interrupt enabled, which is not the case if called from `pm_layered` (https://github.com/RIOT-OS/RIOT/blob/ae437cdd8483f83b3c046a9ec859a66e68bbfe40/sys/pm_layered/pm.c#L61).
The behavior you describe (interrupt being served before re-enabling clocks) is fixed since #7385, but this is not the case anymore.

Reply:
@kYc0o something like a cortexm_wfi() ?

Reply:
I was referring to what @vincent-d just said.

Reply:
But in that case there would be a bunch a irq_disable/restore calls, 3 of each if using pm_layered.